src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/AddUsingsCodeActionResolver.cs

```diff
@@ -0,0 +1,239 @@
+                {
+                    var usingDirectiveLineIndex = codeDocument.Source.Lines.GetLocation(usingDirective.Node.Span.Start).LineIndex;
+                    var head = new Position(usingDirectiveLineIndex, 0);
+                    var edit = new TextEdit() { Range = new Range(head, head), NewText = "" };
+                    do
+                    {
+                        edit.NewText += $"@using {namespaces.Dequeue()}{Environment.NewLine}";
+                    } while (namespaces.Count > 0 && namespaces.Peek().CompareTo(usingDirectiveNamespace) < 0);
+                    edits.Add(edit);
+                }
+            }
+
+            // Add the remaining usings to the end
+            if (namespaces.Count > 0)
+            {
+                var endIndex = usingDirectives.Last().Node.Span.End;
+                var lineIndex = GetLineIndexOrEnd(codeDocument, endIndex - 1) + 1;
+                var head = new Position(lineIndex, 0);
+                var edit = new TextEdit() { Range = new Range(head, head), NewText = "" };
+                do
+                {
+                    edit.NewText += $"@using {namespaces.Dequeue()}{Environment.NewLine}";
+                } while (namespaces.Count > 0);
+                edits.Add(edit);
+            }
+
+            return new WorkspaceEditDocumentChange(new TextDocumentEdit()
+            {
+                TextDocument = codeDocumentIdentifier,
+                Edits = edits,
+            });
+        }
+
+        private static WorkspaceEditDocumentChange GenerateUsingEditsAtTop(
+            RazorCodeDocument codeDocument,
+            VersionedTextDocumentIdentifier codeDocumentIdentifier,
+            string[] namespaceArray)
+        {
+            var namespaceList = namespaceArray.ToList();
+            namespaceList.Sort();
+
+            // If we don't have usings, insert after the last namespace or page directive, which ever comes later
+            var head = new Position(1, 0);
+            var lastNamespaceOrPageDirective = codeDocument.GetSyntaxTree().Root
+                .DescendantNodes()
+                .Where(n => IsNamespaceOrPageDirective(n))
+                .LastOrDefault();
+            if (lastNamespaceOrPageDirective != null)
+            {
+                var lineIndex = GetLineIndexOrEnd(codeDocument, lastNamespaceOrPageDirective.Span.End - 1) + 1;
+                head = new Position(lineIndex, 0);
+            }
+
+            // Insert all usings at the given point
+            var range = new Range(head, head);
+            return new WorkspaceEditDocumentChange(new TextDocumentEdit
+            {
+                TextDocument = codeDocumentIdentifier,
+                Edits = new[]
+                {
+                    new TextEdit()
+                    {
+                        NewText = string.Concat(namespaceList.Select(n => $"@using {n}{Environment.NewLine}")),
```

Comment:
Same confusion about needing multiple namespaces

Reply:
Hmm, this is probably fine for this PR, but is `Environment.NewLine` wrong here? Supposing that the LanguageServer is hosted on Linux but the client is on a Windows dev-box. Also if the rest of the file is using `\r\n` I would expect our actions to use `\r\n` regardless of the systems newline settings. I bet this isn't the only place where we've done this either.

Reply:
Really this just exists for the sake of extensibility. I figured it wouldn't take much time to offer the functionality and there's minimal perf overhead.